cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/KafkaLogProcessorPipeline.java

```diff
@@ -0,0 +1,623 @@
+    return "LogPipeline-" + appender.getName();
+  }
+
+  /**
+   * Returns if the given {@link ILoggingEvent} needs to be appended to the {@link Appender} based on the
+   * log {@link Level}.
+   */
+  private boolean needAppend(ILoggingEvent event) {
+    return event.getLevel().isGreaterOrEqual(effectiveLevelProvider.getEffectiveLevel(event));
+  }
+
+  /**
+   * Initialize offsets for all partitions consumed by this pipeline.
+   *
+   * @param offsets the map for storing the offsets for each partition
+   * @return the same Map passed to this method
+   * @throws InterruptedException if there is an interruption
+   */
+  private Map<Integer, Long> initializeOffsets(Map<Integer, Long> offsets) throws InterruptedException {
+    // Setup initial offsets
+    Set<Integer> partitions = config.getPartitions();
+    while (offsets.size() != partitions.size() && !stopped) {
+      for (int partition : partitions) {
+        Checkpoint checkpoint = checkpoints.get(partition);
+        if (checkpoint != null) {
+          // TODO: (CDAP-7684) Deal with wrong offset issue
+          offsets.put(partition, checkpoint.getNextOffset());
+          continue;
+        }
+
+        try {
+          // If no checkpoint, fetch from the beginning.
+          offsets.put(partition, getLastOffset(partition, kafka.api.OffsetRequest.EarliestTime()));
+        } catch (Exception e) {
+          LOG.info("Failed to get Kafka earliest offset in {}:{} for appender {}. Will be retried",
+                   config.getTopic(), partition, appender.getName());
+          LOG.debug("Failed to get Kafka earliest offset in {}:{} for appender {}.",
+                    config.getTopic(), partition, appender.getName(), e);
+          TimeUnit.SECONDS.sleep(1);
+          break;
+        }
+      }
+    }
+
+    return offsets;
+  }
+
+  /**
+   * Process messages fetched from a given partition.
+   */
+  private long processMessages(String topic, int partition,
+                               Future<Iterable<MessageAndOffset>> future) throws InterruptedException, IOException {
+    Iterable<MessageAndOffset> messages;
+    try {
+      messages = future.get();
+    } catch (ExecutionException e) {
+      try {
+        throw e.getCause();
+      } catch (OffsetOutOfRangeException cause) {
+        // This shouldn't happen under normal situation.
+        // If happened, usually is caused by race between kafka log rotation and fetching in here,
+        // hence just fetching from the beginning should be fine
+        return getLastOffset(partition, kafka.api.OffsetRequest.EarliestTime());
+      } catch (IOException cause) {
+        throw cause;
+      } catch (Throwable t) {
+        // For other type of exceptions, just throw an IOException. It will be handled by caller.
+        throw new IOException(t);
+      }
+    }
+
+    long offset = -1L;
+    for (MessageAndOffset message : messages) {
+      if (eventQueue.getEventSize() >= config.getMaxQueueSize()) {
```

Comment:
it is configurable from the framework size. everything in the `config` object comes from the framework.